### PR TITLE
Added ACM Lab workload

### PR DESCRIPTION
**Pull request by mvazquezc**
##### SUMMARY
Added a new workload to be used by an ACM lab.

##### ISSUE TYPE

- New role Pull Request

##### COMPONENT NAME
ocp4_workload_rhacm

##### ADDITIONAL INFORMATION
The role needs to be executed passing two variables:
* ocp4_workload_rhacm_docker_username
* ocp4_workload_rhacm_docker_password

That user and password must have access to the tech preview repository at this point.

### CODE REVIEW

File: ansible/roles/ocp4-workload-acm-app-policies-lab/files/pull-secret.yaml [hunk 1]
```diff
@@ -0,0 +1,28 @@
+$ANSIBLE_VAULT;1.1;AES256
+62646264666563363963346139366166353734356263373332393966623437363538666437323435
+6236353865626134663532643033613935636565366137640a643937623932366164626561663137
```
Comment (fridim) [inline ansible/roles/ocp4-workload-acm-app-policies-lab/files/pull-secret.yaml:3]: as discussed, can you use a secret variable instead ?

`ocp4_workload_acm_app_policies_lab_pull_secret_content`

That will contain the content of your file, and then use is in the `copy` module

File: ansible/roles/ocp4-workload-acm-app-policies-lab/tasks/remove_workload.yml [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+
+# 1. Ensure we're connected to the cluster
+# 2. Make a temporary dir for cloning the deploy repo
+# 3. Clone the deploy repository
+# 4. Run the uninstall script
+
+
+- name: Ensure there is connectivity to OCP cluster as system:admin
+  command: oc whoami
+  register: whoami
```
Comment (fridim) [inline ansible/roles/ocp4-workload-acm-app-policies-lab/tasks/remove_workload.yml:12]: can add `changed_when: false`

File: ansible/roles/ocp4-workload-acm-app-policies-lab/tasks/remove_workload.yml [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+    path: "deploy"
+    state: absent
+  become: true
+
+- name: Ensure deploy repository is cloned (2/2) - Clone the repository
+  git:
+    repo: "{{ _acm_install_repo }}"
+    dest: "deploy"
+    version: "{{ _acm_install_repo_branch }}"
+
```
Comment (fridim) [inline ansible/roles/ocp4-workload-acm-app-policies-lab/tasks/remove_workload.yml:30]: Can you add and use a `_acm_install_repo_dest` variable for that ?

Do you really need 2 tasks for that ?
Did you try `force` from the git module ?
If you don't need git history, you can use also `depth: 1`


```yaml
- name: Ensure deploy repository is cloned (2/2) - Clone the repository
  git:
    repo: "{{ _acm_install_repo }}"
    dest: "{{ _acm_install_repo_dest }}"
    version: "{{ _acm_install_repo_branch }}"
    force: true
    depth: 1
```

File: ansible/roles/ocp4-workload-acm-app-policies-lab/tasks/workload.yml [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+  file:
+    path: "deploy"
+    state: absent
+  become: true
+
+- name: Ensure deploy repository is cloned (2/3) - Clone the repository
+  git:
+    repo: "{{ _acm_install_repo }}"
+    dest: "deploy"
+    version: "{{ _acm_install_repo_branch }}"
```
Comment (fridim) [inline ansible/roles/ocp4-workload-acm-app-policies-lab/tasks/workload.yml:31]: same comment

File: ansible/roles/ocp4-workload-acm-app-policies-lab/defaults/main.yml [hunk 5]
```diff
@@ -0,0 +1,13 @@
+---
+become_override: False
+ocp_username: opentlc-mgr
+silent: False
+
+_acm_project: "open-cluster-management"
```
Comment (wkulhanek) [inline ansible/roles/ocp4-workload-acm-app-policies-lab/defaults/main.yml:6]: Can we rename these variables to use the workload name as the prefix? And rename the workload to use underscores?

See roles_workload/ocp4_workload_example for how to.

File: ansible/roles/ocp4-workload-acm-app-policies-lab/files/wait-for-deployment [hunk 6]
```diff
@@ -0,0 +1,49 @@
+#!/bin/bash
+
+READY=0
+WAIT=0
+MAX_WAIT=300
+CLUSTER="$1"
+DEPLOYMENT_NAMESPACE="$2"
+DEPLOYMENT_NAME="$3"
+echo "Checking if deployment ${DEPLOYMENT_NAME} from namespace ${DEPLOYMENT_NAMESPACE} on cluster ${CLUSTER} is ready"
```
Comment (wkulhanek) [inline ansible/roles/ocp4-workload-acm-app-policies-lab/files/wait-for-deployment:9]: Can we use k8s means for this instead of shell scripts? Logging workload has examples how to.

Comment (fridim) [reply ansible/roles/ocp4-workload-acm-app-policies-lab/files/wait-for-deployment:9]: I asked the same question on chat (agnosticd room)>
Those scripts are for students during class, not automation

Comment (wkulhanek) [reply ansible/roles/ocp4-workload-acm-app-policies-lab/files/wait-for-deployment:9]: Oh. That wasn’t obvious. :-)

**Comment by mvazquezc** [comment]
@fridim @wkulhanek I just updated the role with your suggestions, could you please review it again?

File: ansible/roles/ocp4_workload_rhacm/tasks/workload.yml [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+    src: "{{ item.src }}"
+    dest: "{{ item.dest }}"
+    mode: 0666
+  loop:
+    - { src: ./templates/subscription.j2, dest: /tmp/subscription.yaml }
+    - { src: ./templates/multiClusterHub.j2, dest: /tmp/multiClusterHub.yaml }
+    - { src: ./templates/operatorGroup.j2, dest: /tmp/operatorGroup.yaml }
+
+- name: Ensure RHACM Namespace is present
+  shell: oc create namespace "{{ ocp4_workload_rhacm_acm_project }}"
```
Comment (fridim) [inline ansible/roles/ocp4_workload_rhacm/tasks/workload.yml:29]: can you use the k8s module instead ?

File: ansible/roles/ocp4_workload_rhacm/tasks/workload.yml [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+
+- name: Ensure RHACM OperatorGroup is present
+  shell: oc -n "{{ ocp4_workload_rhacm_acm_project }}" apply -f /tmp/operatorGroup.yaml
+
+- name: Ensure RHACM Subscription is present
+  shell: oc -n "{{ ocp4_workload_rhacm_acm_project }}" apply -f /tmp/subscription.yaml
```
Comment (fridim) [inline ansible/roles/ocp4_workload_rhacm/tasks/workload.yml:35]: same here

File: ansible/roles/ocp4_workload_rhacm/tasks/workload.yml [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+  command: oc -n "{{ ocp4_workload_rhacm_acm_project }}" patch MultiClusterHub multiclusterhub -p '{"spec":{"forcedReconcile":false}}' --type=merge
+  when: rhacm_multiclusterhub_status.stdout == "Pending"
+
+- name: Wait until RHACM MultiClusterHub is Ready (Second Try)
+  shell: oc -n "{{ ocp4_workload_rhacm_acm_project }}" get MultiClusterHub multiclusterhub -o jsonpath='{.status.phase}'
+  register: rhacm_multiclusterhub_status_2
+  retries: 5
+  delay: 10
+  until: rhacm_multiclusterhub_status_2.stdout == "Running"
+  when: rhacm_multiclusterhub_status.stdout == "Pending"
```
Comment (fridim) [inline ansible/roles/ocp4_workload_rhacm/tasks/workload.yml:72]: Could all those `oc` command be converted to using the k8s module ?

Comment (mvazquezc) [reply ansible/roles/ocp4_workload_rhacm/tasks/workload.yml:72]: Yep, let me change that.

File: ansible/roles/ocp4_workload_rhacm/meta/main.yml [hunk 10]
```diff
@@ -0,0 +1,16 @@
+
+---
+galaxy_info:
+  role_name: ocp4-workload-rhacm
```
Comment (wkulhanek) [inline ansible/roles/ocp4_workload_rhacm/meta/main.yml:4]: Minor nit: role name has now underscores.

File: ansible/roles/ocp4_workload_rhacm/readme.adoc [hunk 11]
```diff
@@ -0,0 +1,124 @@
...
+# a TARGET_HOST is specified in the command line, without using an inventory file
+ansible-playbook -i ${TARGET_HOST}, ./configs/ocp-workloads/ocp-workload.yml \
+    -e"ansible_ssh_private_key_file=~/.ssh/keytoyourhost.pem" \
+    -e"ansible_user=ec2-user" \
+    -e"ocp_username=${OCP_USERNAME}" \
+    -e"ocp_workload=${WORKLOAD}" \
+    -e"guid=${GUID}" \
+    -e"ACTION=remove"
+----
+
```
Comment (wkulhanek) [inline ansible/roles/ocp4_workload_rhacm/readme.adoc:75]: I usually delete everything after that line. It's just not helpful...

File: ansible/roles/ocp4_workload_rhacm/tasks/workload.yml (hunk 7 above)
Comment (wkulhanek) [reply ansible/roles/ocp4_workload_rhacm/tasks/workload.yml:29]: Seconded

File: ansible/roles/ocp4_workload_rhacm/tasks/workload.yml [hunk 12]
```diff
@@ -0,0 +1,78 @@
...
+
+- name: Ensure RHACM OperatorGroup is present
+  shell: oc -n "{{ ocp4_workload_rhacm_acm_project }}" apply -f /tmp/operatorGroup.yaml
+
+- name: Ensure RHACM Subscription is present
+  shell: oc -n "{{ ocp4_workload_rhacm_acm_project }}" apply -f /tmp/subscription.yaml
+
+# Wait for the operator deployment
+- name: Wait until RHACM Subscription is Ready
+  command: oc -n "{{ ocp4_workload_rhacm_acm_project }}" get csv "{{ ocp4_workload_rhacm_acm_csv }}" -o jsonpath='{.status.phase}'
```
Comment (wkulhanek) [inline ansible/roles/ocp4_workload_rhacm/tasks/workload.yml:39]: using `k8s_info` you can do the same a bit more natively. See examples in the serverless, logging or elasticsearch workloads (roles_workloads directgory!)

**Comment by mvazquezc** [comment]
@wkulhanek @fridim updated the playbook, not it uses k8s module.

File: ansible/roles/ocp4_workload_rhacm/tasks/pre_workload.yml [hunk 13]
```diff
@@ -1,5 +1,45 @@
 ---
 # Implement your Pre Workload deployment tasks here
+- set_fact:
+    ansible_python_interpreter: "/bin/python"
```
Comment (fridim) [inline ansible/roles/ocp4_workload_rhacm/tasks/pre_workload.yml:4]: Why is this necessary ?

in rhel8 it's different IIRC, so that would probably not work there.
i'd rather let ansible detect the default python interpreter.

Comment (mvazquezc) [reply ansible/roles/ocp4_workload_rhacm/tasks/pre_workload.yml:4]: If you don't put that, the first run will work, but the next run it will say you need to install python2 bindings for the pip module to work.

File: ansible/roles/ocp4_workload_rhacm/meta/main.yml (hunk 10 above)
Comment (mvazquezc) [reply ansible/roles/ocp4_workload_rhacm/meta/main.yml]: fixed

File: ansible/roles/ocp4_workload_rhacm/readme.adoc (hunk 11 above)
Comment (mvazquezc) [reply ansible/roles/ocp4_workload_rhacm/readme.adoc]: done.

File: ansible/roles/ocp4_workload_rhacm/tasks/workload.yml (hunk 12 above)
Comment (mvazquezc) [reply ansible/roles/ocp4_workload_rhacm/tasks/workload.yml]: done.

File: ansible/roles/ocp4_workload_rhacm/tasks/workload.yml (hunk 9 above)
Comment (mvazquezc) [reply ansible/roles/ocp4_workload_rhacm/tasks/workload.yml]: done.

File: ansible/roles/ocp4_workload_rhacm/tasks/workload.yml (hunk 7 above)
Comment (mvazquezc) [reply ansible/roles/ocp4_workload_rhacm/tasks/workload.yml]: done.

File: ansible/roles/ocp4_workload_rhacm/tasks/workload.yml (hunk 8 above)
Comment (mvazquezc) [reply ansible/roles/ocp4_workload_rhacm/tasks/workload.yml]: done.

File: ansible/roles/ocp4-workload-acm-app-policies-lab/tasks/workload.yml (hunk 4 above)
Comment (mvazquezc) [reply ansible/roles/ocp4-workload-acm-app-policies-lab/tasks/workload.yml]: This is not required anymore.

File: ansible/roles/ocp4-workload-acm-app-policies-lab/files/pull-secret.yaml (hunk 1 above)
Comment (mvazquezc) [reply ansible/roles/ocp4-workload-acm-app-policies-lab/files/pull-secret.yaml]: I moved this to a variable.

File: ansible/roles/ocp4_workload_rhacm/tasks/pre_workload.yml (hunk 13 above)
Comment (wkulhanek) [reply ansible/roles/ocp4_workload_rhacm/tasks/pre_workload.yml:4]: @mvazquezc This should really be done by the workload deployer itself. We have the k8s environment on the hosts. 

E.g. https://github.com/redhat-cop/agnosticd/blob/development/ansible/configs/ocp4-cluster/workloads.yml#L11
or https://github.com/redhat-cop/agnosticd/blob/development/ansible/configs/ocp4-workshop/ocp_workloads.yml#L13

or even: https://github.com/redhat-cop/agnosticd/blob/development/ansible/configs/ocp-workloads/ocp-workload.yml#L11

File: ansible/roles/ocp4_workload_rhacm/tasks/pre_workload.yml [hunk 14]
```diff
@@ -0,0 +1,48 @@
+---
+# Implement your Pre Workload deployment tasks here
+- set_fact:
+    ansible_python_interpreter: "/bin/python"
+
+- name: Setup pre-requisite python3 packages
```
Comment (wkulhanek) [inline ansible/roles/ocp4_workload_rhacm/tasks/pre_workload.yml:6]: I'm not quite sure I understand why you can't use the default k8s virtualenv that's already there for this.

File: ansible/roles/ocp4_workload_rhacm/tasks/pre_workload.yml [hunk 15]
```diff
@@ -0,0 +1,48 @@
...
+- name: Copy requirements.txt to target for k8s virtualenv
+  copy:
+    src: ./files/requirements_k8s.txt
+    dest: /tmp/requirements_k8s.txt
+  become: true
+
+- name: "Create virtualenv k8s"
+  pip:
+    requirements: /tmp/requirements_k8s.txt
+    virtualenv: "/opt/virtualenvs/k8s"
```
Comment (wkulhanek) [inline ansible/roles/ocp4_workload_rhacm/tasks/pre_workload.yml:31]: And this will be a no-op.... except your requirements may override the ones already there.

File: ansible/roles/ocp4_workload_rhacm/tasks/workload.yml [hunk 16]
```diff
@@ -0,0 +1,131 @@
+---
+# Implement your Workload deployment tasks here
+
+- name: Setting up workload for user
+  debug:
+    msg: "Setting up workload for user ocp_username = {{ ocp_username }}"
+
+- set_fact:
```
Comment (wkulhanek) [inline ansible/roles/ocp4_workload_rhacm/tasks/workload.yml:8]: Once again this should not be necessary since it's automatically done by the playbook that calls the workload.

File: ansible/roles/ocp4_workload_rhacm/tasks/pre_workload.yml (hunk 13 above)
Comment (mvazquezc) [reply ansible/roles/ocp4_workload_rhacm/tasks/pre_workload.yml]: I tried with a different lab from opentlc and it seems it works, I cannot fully validate the workload since I need 4.4 cluster. (I was using the installation lab for testing, that's why I needed to setup the virtualenv).

File: ansible/roles/ocp4_workload_rhacm/tasks/pre_workload.yml (hunk 14 above)
Comment (mvazquezc) [reply ansible/roles/ocp4_workload_rhacm/tasks/pre_workload.yml]: Changed.

File: ansible/roles/ocp4_workload_rhacm/tasks/pre_workload.yml (hunk 15 above)
Comment (mvazquezc) [reply ansible/roles/ocp4_workload_rhacm/tasks/pre_workload.yml]: Changed.

File: ansible/roles/ocp4_workload_rhacm/tasks/workload.yml (hunk 16 above)
Comment (mvazquezc) [reply ansible/roles/ocp4_workload_rhacm/tasks/workload.yml]: Changed.

File: ansible/roles/ocp4_workload_rhacm/files/requirements_k8s.txt [hunk 17]
```diff
@@ -0,0 +1,40 @@
+adal==1.2.2
```
Comment (wkulhanek) [inline ansible/roles/ocp4_workload_rhacm/files/requirements_k8s.txt:1]: You don't need that file anymore, do you?

File: ansible/roles/ocp4_workload_rhacm/tasks/pre_workload.yml (hunk 13 above)
Comment (wkulhanek) [reply ansible/roles/ocp4_workload_rhacm/tasks/pre_workload.yml]: Ah. Makes sense.
